File: sorting/merge.py
from heapq import merge
 
def merge_sort(m):
    if len(m) <= 1:
        return m
 
    middle = len(m) / 2
    left = m[:middle]
    right = m[middle:]
 
    left = merge_sort(left)
    right = merge_sort(right)
    return list(merge(left, right))

def merge(left, right):
    result = []
 
    while left and right:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
 
    if left:
        result.extend(left)
    if right:
        result.extend(right)
    return result

File: sorting/ruby/merge.rb
#!/bin/usr/ruby
# Works with Array 
# ary.mergesort {|a, b| b <=> a}
 
class Array
  def mergesort(&comparitor)
    if length <= 1
      self
    else
      unless comparitor
        comparitor = lambda {|a, b| a <=> b}
      end
      middle = length / 2
      left  = self[0,  middle].mergesort(&comparitor)
      right = self[middle..-1].mergesort(&comparitor)
      merge(left, right, comparitor) 
    end
  end
 
  protected 
  def merge(left, right, comparitor)
    if left.empty?
      right
    elsif right.empty?
      left
    elsif comparitor.call(left.first, right.first) <= 0
      [left.first] + merge(left[1..-1], right, comparitor)
    else
      [right.first] + merge(left, right[1..-1], comparitor)
    end
  end
end
